homeassistant/components/trafikverket_train/sensor.py

```diff
@@ -0,0 +1,147 @@
+    vol.Required(CONF_TRAINS): [{
+        vol.Required(CONF_NAME): cv.string,
+        vol.Required(CONF_TO): cv.string,
+        vol.Required(CONF_FROM): cv.string,
+        vol.Optional(CONF_TIME): cv.time}]
+})
+
+
+@asyncio.coroutine
+def async_setup_platform(
+        hass, config, async_add_devices, discovery_info=None):
+    """Set up the departure sensor."""
+    _LOGGER.debug("Start async_setup_platform")
+    from pytrafikverket import TrafikverketTrain
+    httpsession = async_get_clientsession(hass)
+    train_api = TrafikverketTrain(httpsession, config.get(CONF_API_KEY))
+    sensors = []
+    for train in config.get(CONF_TRAINS):
+        from_sig = yield from train_api.async_get_train_station(
+            train.get(CONF_FROM))
+        to_sig = yield from train_api.async_get_train_station(
+            train.get(CONF_TO))
+        sensor = TrainSensor(train_api,
+                             train.get(CONF_NAME),
+                             from_sig,
+                             to_sig,
+                             train.get(CONF_TIME))
+        sensors.append(sensor)
+
+    async_add_devices(sensors, update_before_add=True)
+    _LOGGER.debug("end async_setup_platform")
+
+
+class TrainSensor(Entity):
+    """Contains data about a train depature."""
+
+    def __init__(self, train_api, name, from_sig, to_sig, time):
+        """Initialize the sensor."""
+        self._train_api = train_api
+        self._name = name
+        self._from_sig = from_sig
+        self._to_sig = to_sig
+        self._time = time
+        self._state = None
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Retrieve latest state."""
+        _LOGGER.debug("start async_update")
+        if self._time is not None:
+            when = datetime.combine(date.today(), self._time)
+            _LOGGER.debug("self._from_sig: %s, self._to_sig: %s, when: %s",
+                          self._from_sig.name, self._to_sig.name, when)
+            try:
+                self._state = yield from \
+                    self._train_api.async_get_train_stop(
+                        self._from_sig, self._to_sig, when)
+            except ValueError:
+                _LOGGER.debug("API error or no departure found for time %s",
+                              when)
+        else:
+            when = datetime.now()
+            _LOGGER.debug("self._from_sig: %s, self._to_sig: %s, when: %s",
+                          self._from_sig.name, self._to_sig.name, when)
+            self._state = yield from \
+                self._train_api.async_get_next_train_stop(
+                    self._from_sig, self._to_sig, when)
+        _LOGGER.debug("end async_update")
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        if self._state is None:
+            return None
+        state = self._state
+        other_information = None
+        if state.other_information is not None and state.other_information:
```

Comment:
Looks like that the second check is superfluous.

Reply:
Corrected this one and the next one that was similar. 
Tested and works fine without the second check in the different scenarios.